### PR TITLE
[MetaSchedule] Enable Clone Function for Task-Level Classes

### PR DESCRIPTION
**Pull request by zxybazh**
This PR introduces a clone function for each of the task-level MetaSchedule classes for convenient class deep copying.

- [x] ScheduleRule
- [x] Postproc
- [x] Mutator
- [x] SpaceGenerator
- [x] SearchStrategy
- [x] TuneContext

cc @Hzfengsy @junrushao1994

### CODE REVIEW

File: src/meta_schedule/schedule_rule/add_rfactor.cc [hunk 1]
```diff
@@ -36,6 +36,16 @@ class AddRFactorNode : public ScheduleRuleNode {
   // Inherited from ScheduleRuleNode
   Array<tir::Schedule> Apply(const tir::Schedule& sch, const tir::BlockRV& block_rv);
 
+  // Inherited from ScheduleRuleNode
+  ScheduleRule Clone() final {
+    ObjectPtr<AddRFactorNode> n = make_object<AddRFactorNode>(*this);
+    n->max_jobs_per_core = this->max_jobs_per_core;
+    n->max_innermost_factor = this->max_innermost_factor;
+    n->max_parallel_extent_ = this->max_parallel_extent_;
+    n->max_parallel_basic_ = this->max_parallel_basic_;
```
Comment (junrushao) [inline src/meta_schedule/schedule_rule/add_rfactor.cc:45]: Those fields are already copied in `make_object<AddRFactorNode>(*this)`

**Comment by junrushao** [comment]
This is used for enhancing UX of the tuning. Considering the current `tune` API:
https://github.com/apache/tvm/blob/370abe69d24519a5453cead846d328a1c378957f/tests/python/integration/test_meta_schedule_auto_tensorize.py#L174-L175
where users have to provide a lambda function, has become a bit confusing and inconvenient, we want to make it possible for users to directly provide a single object, and our system takes care of cloning

File: src/meta_schedule/tune_context.cc [hunk 2]
```diff
@@ -52,6 +52,32 @@ TuneContext::TuneContext(Optional<IRModule> mod,
...
+  }
+  if (this->mutator_probs.defined()) {
+    n->mutator_probs = Map<Mutator, FloatImm>();
+    for (const auto& kv : this->mutator_probs) {
+      n->mutator_probs.Set(kv.first->Clone(), kv.second);
+    }
+  }
+  if (this->space_generator.defined()) n->space_generator = this->space_generator.value()->Clone();
+  if (this->search_strategy.defined()) n->search_strategy = this->search_strategy.value()->Clone();
+  n->Initialize();
```
Comment (junrushao) [inline src/meta_schedule/tune_context.cc:77]: Let's fork the random state instead of copying here

**Review by junrushao — APPROVED** [review]
Overall LGTM! BTW, please address my comments in a separate PR

File: src/meta_schedule/tune_context.cc (hunk 2 above)
Comment (zxybazh) [reply src/meta_schedule/tune_context.cc:77]: Sounds good to me!

**Pull request merged**